README.md
# JavaScript 方法库
## github地址
```
https://github.com/detanx/js-self-methods
```
## vscode插件
1. 安装vscode中插件搜索框输入**js-self-methods**
2. 安装完毕后按**F1**(Win:ctrl+shift+p、Mac:command+shift+p)
3. 输入"**method**"可以查看方法列表.
4. 或者**输入关键词**, 比如**"防抖"**.
### 方法列表
* method: 节流
* method: 防抖
* method: 柯里化
* method: 斐波拉契数列
* method: 记忆函数
* method: 金额千位分隔符
* method: 敏感符号转义


CHANGELOG.md
# Change Log

All notable changes to the "jsselfmethods" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/js/extend.js
/*
 * @Author: xudong.tang(Detanx)
 * @Date: 2019-08-15 14:44:24
 * @LastEditors: xudong.tang(Detanx)
 * @LastEditTime: 2019-08-23 14:10:13
 * @Email: detanxit@163.com;detanxit163@gmail.com
 * @Description: 常用的一些方法
 */
const extend = () => {
    // ********-- 节流 --********
    function throttle(fn, delay = 500) {
        let flag = true;
        return function (...args) {
            if (!flag) return;
            flag = false;
            setTimeout(() => {
                fn.apply(this, args);
                flag = true;
            }, delay);
        };
    }

    // ********-- 防抖 --********
    function debounce (fn, delay = 500)  {
        let timer = null
        return function (...args) {
            if (timer) {
                clearTimeout(timer)
            }
            timer = setTimeout(() => {
                fn(...args)
            }, delay)

        }
    }

    // ********-- 柯里化 --********
    function _selfCurry  (fn, args = []) {
        const fnLen = fn.length
        return function () {
            // newArgs = args.concat(Array.prototype.slice.call(arguments))
            // newArgs = args.concat([],[...arguments])
            const newArgs = [...args, ...arguments]
            const newArgsLen = newArgs.length
            if (newArgsLen < fnLen) {
                return _selfCurry.call(this, fn, newArgs)
                // return _selfCurry.apply(this,[fn,newArgs])
            }
            return fn.call(this, ...newArgs)
            // return fn.apply(this, newArgs)

        }
    }

    // ********-- 斐波拉契数列 --********
    function febonacci (n) {
        if (n < 1) throw new Error('Param is error')
        if (n === 1 || n === 2) return 1
        return febonacci(n - 1) + febonacci(n - 2)
    }

    // ********-- 记忆函数 --********
    function memory (fn) {
        let obj = {}
        return function (n) {
            if (obj[n] === undefined) obj[n] = fn(n)
            return obj[n]
        }
    }

    // ********-- 金额千位分隔符 --********
    function moneyFormat (num)  {
        if (Object.prototype.toString.call(num) !== '[object Number]') {
            throw new SyntaxError('Param not a number')
        }
        num = parseFloat(num.toFixed(3));
        let [integer, decimal] = String.prototype.split.call(num, '.');
        integer = integer.replace(/\d(?=(\d{3})+$)/g, '$&,');
        return integer + '.' + (decimal ? decimal : '');
    }
    // ********-- 敏感符号转义 --********
    function escape  (str) {
        const type = {
            '"': '&quot;',
            '&': '&amp;',
            '<': '&lt;',
            '>': '&gt;'
        }
        return str.replace(/["<>&]/g, char => {
            return type[char]
        })
    }
    return {
        moneyFormat,
        _selfCurry,
        throttle,
        debounce,
        febonacci,
        memory,
        escape
    }
}
module.exports = extend()

scripts/readme.js
/*
 * @Author: xudong.tang(Detanx)
 * @Date: 2019-08-23 11:18:07
 * @LastEditors: xudong.tang(Detanx)
 * @LastEditTime: 2019-08-28 11:00:49
 * @Email: detanxit@163.com;detanxit163@gmail.com
 * @Description: 
 */
const fs = require('fs');
const chalk = require('chalk');
const json = require('../src/js/methods');
let content = `# JavaScript 方法库
## github地址
\`\`\`
https://github.com/detanx/js-self-methods
\`\`\`
## vscode插件
1. 安装vscode中插件搜索框输入**js-self-methods**
2. 安装完毕后按**F1**(Win:ctrl+shift+p、Mac:command+shift+p)
3. 输入"**method**"可以查看方法列表.
4. 或者**输入关键词**, 比如**"防抖"**.
### 方法列表
`;

json.forEach(({title, fun})=>{
    content+= `* method: ${title}\n`;
    // content+= `\`\`\`javascript\r\n`;
    // content+= `${fun}\r\n\`\`\`\r\n`;
});
fs.writeFileSync('./README.md',content,'utf-8');
console.log(chalk.green('readme构建完成'));

src/js/methods.js
/*
 * @Author: xudong.tang(Detanx)
 * @Date: 2019-08-23 11:13:31
 * @LastEditors: xudong.tang(Detanx)
 * @LastEditTime: 2019-08-23 14:16:43
 * @Email: detanxit@163.com;detanxit163@gmail.com
 * @Description: 
 */
const extend = require('./extend')

module.exports = [
    {
        title: '节流',
        name: 'throttle',
        fun: extend.throttle
    },
    {
        title: '防抖',
        name: 'debounce',
        fun: extend.debounce
    },
    {
        title: '柯里化',
        name: '_selfCurry',
        fun: extend._selfCurry
    },
    {
        title: '斐波拉契数列',
        name: 'febonacci',
        fun: extend.febonacci
    },
    {
        title: '记忆函数',
        name: 'memory',
        fun: extend.memory
    },
    {
        title: '金额千位分隔符',
        name: 'moneyFormat',
        fun: extend.moneyFormat
    },
    {
        title: '敏感符号转义',
        name: 'escape',
        fun: extend.escape
    },
];

src/extension.ts
/*
 * @Author: xudong.tang(Detanx)
 * @Date: 2019-08-23 10:51:41
 * @LastEditors: xudong.tang(Detanx)
 * @LastEditTime: 2019-08-23 16:08:06
 * @Email: detanxit@163.com;detanxit163@gmail.com
 * @Description: 
 */
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';

import { IMethod } from './interface'
const METHODS = require('../src/js/methods.js');
// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {

	// Use the console to output diagnostic information (console.log) and errors (console.error)
	// This line of code will only be executed once when your extension is activated
	console.log('Congratulations, your extension "jsselfmethods" is now active!');

	METHODS.forEach(({ fun, name }: IMethod) => {
		let disposable = vscode.commands.registerCommand(`extension.${name}`, () => {
			// The code you place here will be executed every time your command is executed
			const editor = vscode.window.activeTextEditor;
			if (editor) {
				const { selections } = editor;

				editor.edit(editBuilder => {
					selections.forEach(selection => {
						const { start, end } = selection;
						const range = new vscode.Range(start, end);
						editBuilder.replace(range, String(fun));
					});
				});
				// Display a message box to the user
				vscode.window.showInformationMessage(`已插入方法: ${name}`);
			} else {
				vscode.window.showWarningMessage('js-self-methods: 只有编辑文本的时候才可以使用!');
			}
		});
		context.subscriptions.push(disposable);
	});

}

// this method is called when your extension is deactivated
export function deactivate() {}
